src/main/java/com/google/devtools/build/lib/skyframe/PackageFunction.java

```diff
@@ -1067,99 +1070,100 @@ private LoadedPackageCacheEntry loadPackage(
-                packageId,
+        input =
+            ParserInputSource.create(
+                FileSystemUtils.convertFromLatin1(buildFileBytes), inputFile.asFragment());
+        StoredEventHandler astParsingEventHandler = new StoredEventHandler();
+        BuildFileAST ast =
+            PackageFactory.parseBuildFile(
+                packageId, input, preludeStatements, repositoryMapping, astParsingEventHandler);
+        astParseResult = new AstParseResult(ast, astParsingEventHandler);
+        astCache.put(packageId, astParseResult);
+      }
+      SkylarkImportResult importResult;
+      try {
+        importResult =
+            fetchImportsFromBuildFile(
                 buildFilePath,
-                astParseResult,
-                importResult.importMap,
-                importResult.fileDependencies,
-                defaultVisibility,
-                starlarkSemantics,
-                globberWithSkyframeGlobDeps);
-        long loadTimeNanos = Math.max(BlazeClock.nanoTime() - startTimeNanos, 0L);
-        packageCacheEntry = new LoadedPackageCacheEntry(
-            pkgBuilder,
-            globberWithSkyframeGlobDeps.getGlobDepsRequested(),
-            loadTimeNanos);
-        numPackagesLoaded.incrementAndGet();
-        if (packageProgress != null) {
-          packageProgress.doneReadPackage(packageId);
-        }
-        packageFunctionCache.put(packageId, packageCacheEntry);
+                packageId,
+                astParseResult.ast,
+                /* workspaceChunk = */ -1,
+                env,
+                skylarkImportLookupFunctionForInlining);
+      } catch (NoSuchPackageException e) {
+        throw new PackageFunctionException(e, Transience.PERSISTENT);
+      } catch (InterruptedException e) {
+        astCache.invalidate(packageId);
+        throw e;
+      }
+      if (importResult == null) {
+        return null;
+      }
+      astCache.invalidate(packageId);
+      GlobberWithSkyframeGlobDeps globberWithSkyframeGlobDeps =
+          makeGlobber(inputFile, packageId, packageRoot, env);
+      long startTimeNanos = BlazeClock.nanoTime();
+      Package.Builder pkgBuilder =
+          packageFactory.createPackageFromAst(
+              workspaceName,
+              repositoryMapping,
+              packageId,
+              buildFilePath,
+              astParseResult,
+              importResult.importMap,
+              importResult.fileDependencies,
+              defaultVisibility,
+              starlarkSemantics,
+              globberWithSkyframeGlobDeps);
+      long loadTimeNanos = Math.max(BlazeClock.nanoTime() - startTimeNanos, 0L);
+      LoadedPackageCacheEntry packageCacheEntry =
+          new LoadedPackageCacheEntry(
+              pkgBuilder, globberWithSkyframeGlobDeps.getGlobDepsRequested(), loadTimeNanos);
+      numPackagesLoaded.incrementAndGet();
+      packageFunctionCache.put(packageId, packageCacheEntry);
+      if (packageProgress != null) {
```

Comment:
curious, any particular reason you changed the order here, inserting into the cache before emitting the progress message?

In other words, could you simplify this by putting the progress message in a finally block and avoiding the catch clause?

Reply:
We actually don't always want to report the package as done when leaving this function because we might be returning `null` due to a skyframe restart. Skyframe restarts can cause a package to be reported as starting to load multiple times, which is harmless for the current receiver. But reporting a package as done multiple times would artificially inflate the loaded packages count.

Reply:
Ah, of course. Thanks.